### ISSUE TITLE
Exit transition not working

### ISSUE THREAD

**Issue by essarn** [raamcosta/compose-destinations#74]
See the following destinations with the default animations set below. The exit transition does not run and instead the enter transition runs every time. The animations are from [here](https://github.com/fornewid/material-motion-compose/tree/main/core).

```kotlin
// MainActivity.kt

@OptIn(
    ExperimentalAnimationApi::class,
    ExperimentalMaterialNavigationApi::class
)
@Composable
fun Content() {
    val engine = rememberAnimatedNavHostEngine(
        rootDefaultAnimations = RootNavGraphDefaultAnimations(
            enterTransition = { materialFadeThroughIn() },
            exitTransition = { materialFadeThroughOut() }
        )
    )

    DestinationsNavHost(navGraph = NavGraphs.root, engine = engine)
}
```

```kotlin
// HomeScreen.kt

@OptIn(ExperimentalMaterial3Api::class)
@Destination(start = true)
@Composable
fun HomeScreen(navigator: DestinationsNavigator) {
    Scaffold(topBar = { SmallTopAppBar(title = { Text(text = "Home Screen") }) }) {
        Column {
            Text(text = "Hello World!")
            ElevatedButton(onClick = { navigator.navigate(AboutScreenDestination) }) {
                Text(text = "Navigate to About Screen")
            }
        }
    }
}
```

```kotlin
// AboutScreen.kt

@OptIn(ExperimentalMaterial3Api::class)
@Destination
@Composable
fun AboutScreen(navigator: DestinationsNavigator) {
    Scaffold(topBar = {
        SmallTopAppBar(
            title = { Text(text = "About Screen") },
            navigationIcon = {
                IconButton(onClick = { navigator.popBackStack() }) {
                    Icon(Icons.Filled.ArrowBack, contentDescription = null)
                }
            })
    }) {
        Column {
            Text(text = "About Screen!")
        }
    }
}
```

![TWc6dvc](https://user-images.githubusercontent.com/16906089/159376912-148520c9-9efd-4863-8ff8-300c3397385d.gif)

**Comment by raamcosta** [comment]
Hi @essarn !

tbh I haven’t had much time to play with animations yet, I just exposes Accompanist apis through my own. That said, I believe you should look at `popEnterTransition` and `popExitTransition`, since when you click that back arrow button, you are popping and not exiting. Exiting should be applied to your Home Screen when you navigate from Home to About though, I can’t be sure looking at that video if it is or not.

Can you please let me know if this helps you? I’m really confident that if there is an issue it is not with Compose Destinations, since I pretty much just call Accompanist APIs. 
But I am of course keen to help you to help me get to the bottom of this to be sure 🙂

**Comment by essarn** [comment]
Hello @raamcosta 

By default, the `pop` transitions should be the normal ones according to the docs (https://google.github.io/accompanist/navigation-animation/#usage) so that would not be the problem. 

I have a bit of a hard time understanding the order in which the transitions run. What happens when there is both an exit transition for the `initial` destination and an enter transition for the `target` destination? In my tests, it seems that the exit transition never runs. 

For example; in this case, only the enter transition runs when navigating to the `AboutScreen` but not the exit transition when navigation back. I.e there is no fade out when navigating back.

```kotlin
@OptIn(ExperimentalMaterial3Api::class)
@Destination(start = true)
@Composable
fun HomeScreen(navigator: DestinationsNavigator) {
    Scaffold(topBar = { SmallTopAppBar(title = { Text(text = "Home Screen") }) }) {
        Column {
            Text(text = "Hello World!")
            ElevatedButton(onClick = { navigator.navigate(AboutScreenDestination) }) {
                Text(text = "Navigate to About Screen")
            }
        }
    }
}
```

```kotlin
@OptIn(ExperimentalMaterial3Api::class)
@Destination(style = AboutDestinationTransitions::class)
@Composable
fun AboutScreen(navigator: DestinationsNavigator) {
    Scaffold(topBar = {
        SmallTopAppBar(
            title = { Text(text = "About Screen") },
            navigationIcon = {
                IconButton(onClick = { navigator.popBackStack() }) {
                    Icon(Icons.Filled.ArrowBack, contentDescription = null)
                }
            })
    }) {
        Column {
            Text(text = "About Screen!")
        }
    }
}

@OptIn(ExperimentalAnimationApi::class)
object AboutDestinationTransitions : DestinationStyle.Animated {
    override fun AnimatedContentScope<NavBackStackEntry>.enterTransition() = materialFadeThroughIn()

    override fun AnimatedContentScope<NavBackStackEntry>.exitTransition() = fadeOut()
}
```

**Comment by raamcosta** [comment]
Hey again @essarn !

> By default, the pop transitions should be the normal ones according to the docs (https://google.github.io/accompanist/navigation-animation/#usage) so that would not be the problem.

You are right! I was not remembering this was the case, we can even check the constructor of `RootNavGraphDefaultAnimations`:

```kotlin
class RootNavGraphDefaultAnimations(
    override val enterTransition: DestinationEnterTransition = DestinationEnterTransition { EnterTransition.None },
    override val exitTransition: DestinationExitTransition = DestinationExitTransition { ExitTransition.None },
    override val popEnterTransition: DestinationEnterTransition = enterTransition,
    override val popExitTransition: DestinationExitTransition = exitTransition,
): NavGraphDefaultAnimationParams
```

And this explains why the video you posted shows the behavior where whether you are navigating into a screen or popping into one, the animation is the same: the screen "exiting" kinda translates a bit and fades and the screen "entering" translates in the other direction and fades in. At least this is what I can see. Now, I don't know what the actual animations do, but it feels correct to me, now looking at it again.

Basically what happens in your first setup is that both screens have the same enter and exit animations, so:

- When you navigate from HomeScreen to AboutScreen:
    - HomeScreen's exit transition runs
    - AboutScreen's enter transition runs
- When you navigate back to HomeScreen:
    - AboutScreen's pop exit (same as normal exit in your case) transition runs
    - HomeScreen's pop enter (same as normal enter in your case)  transition runs
    
I just tested your second setup, and the fadeOut does run, however, the HomeScreen doesn't have any animations set, so the screen immediatelly shows up above the AboutScreen and the fading doesn't actually show for a long time, but if you set a breakpoint in the `fadeOut` call, you'll see that it runs when you exit the AboutScreen.

**Comment by essarn** [comment]
Thank you for all the help @raamcosta!

What you wrote makes sense and it works now. Thanks again 😄

**Comment by raamcosta** [comment]
Glad I could help! 🙂
